File: src/app_store_connect/api/categories.py
"""
App Categories API module for App Store Connect
Manages primary and secondary app categories
"""

from typing import Dict, Any, List, Optional
from ..base import BaseAPI


class CategoriesAPI(BaseAPI):
    """
    Manage app categories and subcategories in App Store Connect
    """
    
    # Available App Store Categories
    # Reference: https://developer.apple.com/app-store/categories/
    CATEGORIES = {
        'BOOKS': 'Books',
        'BUSINESS': 'Business', 
        'DEVELOPER_TOOLS': 'Developer Tools',
        'EDUCATION': 'Education',
        'ENTERTAINMENT': 'Entertainment',
        'FINANCE': 'Finance',
        'FOOD_AND_DRINK': 'Food & Drink',
        'GAMES': 'Games',
        'GRAPHICS_AND_DESIGN': 'Graphics & Design',
        'HEALTH_AND_FITNESS': 'Health & Fitness',
        'LIFESTYLE': 'Lifestyle',
        'MAGAZINES_AND_NEWSPAPERS': 'Magazines & Newspapers',
        'MEDICAL': 'Medical',
        'MUSIC': 'Music',
        'NAVIGATION': 'Navigation',
        'NEWS': 'News',
        'PHOTO_AND_VIDEO': 'Photo & Video',
        'PRODUCTIVITY': 'Productivity',
        'REFERENCE': 'Reference',
        'SHOPPING': 'Shopping',
        'SOCIAL_NETWORKING': 'Social Networking',
        'SPORTS': 'Sports',
        'TRAVEL': 'Travel',
        'UTILITIES': 'Utilities',
        'WEATHER': 'Weather'
    }
    
    # Game Subcategories
    GAME_SUBCATEGORIES = {
        'ACTION': 'Action',
        'ADVENTURE': 'Adventure',
        'ARCADE': 'Arcade',
        'BOARD': 'Board',
        'CARD': 'Card',
        'CASINO': 'Casino',
        'CASUAL': 'Casual',
        'DICE': 'Dice',
        'EDUCATIONAL': 'Educational',
        'FAMILY': 'Family',
        'MUSIC': 'Music',
        'PUZZLE': 'Puzzle',
        'RACING': 'Racing',
        'ROLE_PLAYING': 'Role Playing',
        'SIMULATION': 'Simulation',
        'SPORTS': 'Sports',
        'STRATEGY': 'Strategy',
        'TRIVIA': 'Trivia',
        'WORD': 'Word'
    }
    
    # Magazines & Newspapers Subcategories
    NEWSSTAND_SUBCATEGORIES = {
        'ARTS_AND_PHOTOGRAPHY': 'Arts & Photography',
        'AUTOMOTIVE': 'Automotive',
        'BRIDES_AND_WEDDINGS': 'Brides & Weddings',
        'BUSINESS_AND_INVESTING': 'Business & Investing',
        'CHILDRENS_MAGAZINES': "Children's Magazines",
        'COMPUTERS_AND_INTERNET': 'Computers & Internet',
        'COOKING_FOOD_AND_DRINK': 'Cooking, Food & Drink',
        'CRAFTS_AND_HOBBIES': 'Crafts & Hobbies',
        'ELECTRONICS_AND_AUDIO': 'Electronics & Audio',
        'ENTERTAINMENT': 'Entertainment',
        'FASHION_AND_STYLE': 'Fashion & Style',
        'HEALTH_MIND_AND_BODY': 'Health, Mind & Body',
        'HISTORY': 'History',
        'HOME_AND_GARDEN': 'Home & Garden',
        'LITERARY_MAGAZINES_AND_JOURNALS': 'Literary Magazines & Journals',
        'MENS_INTEREST': "Men's Interest",
        'MOVIES_AND_MUSIC': 'Movies & Music',
        'NEWS_AND_POLITICS': 'News & Politics',
        'OUTDOORS_AND_NATURE': 'Outdoors & Nature',
        'PARENTING_AND_FAMILY': 'Parenting & Family',
        'PETS': 'Pets',
        'PROFESSIONAL_AND_TRADE': 'Professional & Trade',
        'REGIONAL_NEWS': 'Regional News',
        'SCIENCE': 'Science',
        'SPORTS_AND_LEISURE': 'Sports & Leisure',
        'TEENS': 'Teens',
        'TRAVEL_AND_REGIONAL': 'Travel & Regional',
        'WOMENS_INTEREST': "Women's Interest"
    }
    
    def get_app_categories(self, app_info_id: str) -> Dict[str, Any]:
        """
        Get the current primary and secondary categories for an app
        
        Args:
            app_info_id: The app info ID
            
        Returns:
            Dictionary containing primary and secondary category information
        """
        endpoint = f'appInfos/{app_info_id}'
        params = {
            'fields[appInfos]': 'primaryCategory,secondaryCategory,primarySubcategoryOne,primarySubcategoryTwo,secondarySubcategoryOne,secondarySubcategoryTwo',
            'include': 'primaryCategory,secondaryCategory'
        }
        
        response = super().get(endpoint, params=params)
        app_info = response.get('data', {})
        
        # Extract category relationships
        relationships = app_info.get('relationships', {})
        included = response.get('included', [])
        
        # Build category lookup from included data
        category_lookup = {}
        for item in included:
            if item.get('type') == 'appCategories':
                category_lookup[item['id']] = item['attributes']
        
        # Extract current categories
        result = {
            'primaryCategory': None,
            'secondaryCategory': None,
            'primarySubcategoryOne': None,
            'primarySubcategoryTwo': None,
            'secondarySubcategoryOne': None,
            'secondarySubcategoryTwo': None
        }
        
        # Get primary category
        primary_cat = relationships.get('primaryCategory', {}).get('data')
        if primary_cat:
            result['primaryCategory'] = {
                'id': primary_cat['id'],
                'attributes': category_lookup.get(primary_cat['id'], {})
            }
        
        # Get secondary category
        secondary_cat = relationships.get('secondaryCategory', {}).get('data')
        if secondary_cat:
            result['secondaryCategory'] = {
                'id': secondary_cat['id'],
                'attributes': category_lookup.get(secondary_cat['id'], {})
            }
        
        # Get subcategories from attributes
        attributes = app_info.get('attributes', {})
        result['primarySubcategoryOne'] = attributes.get('primarySubcategoryOne')
        result['primarySubcategoryTwo'] = attributes.get('primarySubcategoryTwo')
        result['secondarySubcategoryOne'] = attributes.get('secondarySubcategoryOne')
        result['secondarySubcategoryTwo'] = attributes.get('secondarySubcategoryTwo')
        
        return result
    
    def update_app_categories(
        self,
        app_info_id: str,
        primary_category_id: Optional[str] = None,
        secondary_category_id: Optional[str] = None,
        primary_subcategory_one: Optional[str] = None,
        primary_subcategory_two: Optional[str] = None,
        secondary_subcategory_one: Optional[str] = None,
        secondary_subcategory_two: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Update app categories and subcategories
        
        Args:
            app_info_id: The app info ID
            primary_category_id: Primary category ID (use get_all_categories to find IDs)
            secondary_category_id: Optional secondary category ID
            primary_subcategory_one: First subcategory for primary category (for Games/Newsstand)
            primary_subcategory_two: Second subcategory for primary category (for Games/Newsstand)
            secondary_subcategory_one: First subcategory for secondary category
            secondary_subcategory_two: Second subcategory for secondary category
            
        Returns:
            Updated app info data
            
        Note:
            - Primary category is required for all apps
            - Secondary category is optional
            - Subcategories are only applicable for Games and Magazines & Newspapers categories
            - You can have up to 2 subcategories per category
        """
        data = {
            'data': {
                'type': 'appInfos',
                'id': app_info_id,
                'attributes': {},
                'relationships': {}
            }
        }
        
        # Add subcategories to attributes if provided
        if primary_subcategory_one is not None:
            data['data']['attributes']['primarySubcategoryOne'] = primary_subcategory_one
        if primary_subcategory_two is not None:
            data['data']['attributes']['primarySubcategoryTwo'] = primary_subcategory_two
        if secondary_subcategory_one is not None:
            data['data']['attributes']['secondarySubcategoryOne'] = secondary_subcategory_one
        if secondary_subcategory_two is not None:
            data['data']['attributes']['secondarySubcategoryTwo'] = secondary_subcategory_two
        
        # Add category relationships if provided
        if primary_category_id:
            data['data']['relationships']['primaryCategory'] = {
                'data': {
                    'type': 'appCategories',
                    'id': primary_category_id
                }
            }
        
        if secondary_category_id:
            data['data']['relationships']['secondaryCategory'] = {
                'data': {
                    'type': 'appCategories',
                    'id': secondary_category_id
                }
            }
        
        response = super().patch(f'appInfos/{app_info_id}', data=data)
        return response.get('data', {})
    
    def get_all_categories(self, platform: str = 'IOS') -> List[Dict[str, Any]]:
        """
        Get all available app categories
        
        Args:
            platform: Platform filter - IOS, MAC_OS, TV_OS (default: IOS)
            
        Returns:
            List of available categories with their IDs and attributes
        """
        endpoint = 'appCategories'
        params = {
            'filter[platforms]': platform,
            'limit': 200  # Get all categories
        }
        
        response = super().get(endpoint, params=params)
        return response.get('data', [])
    
    def get_category_by_name(self, category_name: str, platform: str = 'IOS') -> Optional[Dict[str, Any]]:
        """
        Find a category by its display name
        
        Args:
            category_name: The category display name (e.g., 'Photo & Video', 'Games')
            platform: Platform filter - IOS, MAC_OS, TV_OS (default: IOS)
            
        Returns:
            Category data if found, None otherwise
        """
        categories = self.get_all_categories(platform)
        
        for category in categories:
            # Check against the display name in attributes
            if category.get('attributes', {}).get('displayName') == category_name:
                return category
        
        return None
    
    def set_photo_video_category(self, app_info_id: str) -> Dict[str, Any]:
        """
        Convenience method to set app category to Photo & Video
        Useful for photo editing apps like CleanShot
        
        Args:
            app_info_id: The app info ID
            
        Returns:
            Updated app info data
        """
        # Photo & Video category ID is PHOTO_AND_VIDEO
        return self.update_app_categories(
            app_info_id=app_info_id,
            primary_category_id='PHOTO_AND_VIDEO'
        )
    
    def set_game_category(
        self,
        app_info_id: str,
        game_subcategory_one: str,
        game_subcategory_two: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Convenience method to set app as a game with subcategories
        
        Args:
            app_info_id: The app info ID
            game_subcategory_one: First game subcategory (from GAME_SUBCATEGORIES)
            game_subcategory_two: Optional second game subcategory
            
        Returns:
            Updated app info data
        """
        # Find Games category
        games_category = self.get_category_by_name('Games')
        if not games_category:
            raise ValueError("Games category not found")
        
        # Validate subcategories
        if game_subcategory_one not in self.GAME_SUBCATEGORIES:
            raise ValueError(f"Invalid game subcategory: {game_subcategory_one}")
        
        if game_subcategory_two and game_subcategory_two not in self.GAME_SUBCATEGORIES:
            raise ValueError(f"Invalid game subcategory: {game_subcategory_two}")
        
        return self.update_app_categories(
            app_info_id=app_info_id,
            primary_category_id=games_category['id'],
            primary_subcategory_one=game_subcategory_one,
            primary_subcategory_two=game_subcategory_two
        )

File: rust/src/api/categories.rs
use crate::base::BaseAPI;
use crate::error::AppStoreConnectError;
use serde_json::{json, Value};
use std::collections::HashMap;

#[derive(Clone)]
pub struct CategoriesAPI {
    base: BaseAPI,
}

impl CategoriesAPI {
    pub fn new(base: BaseAPI) -> Self {
        Self { base }
    }

    pub async fn get_app_categories(
        &self,
        app_info_id: &str,
    ) -> Result<Value, AppStoreConnectError> {
        let mut params = HashMap::new();
        params.insert(
            "fields[appInfos]".to_string(),
            "primaryCategory,secondaryCategory,primarySubcategoryOne,primarySubcategoryTwo,secondarySubcategoryOne,secondarySubcategoryTwo".to_string(),
        );
        params.insert(
            "include".to_string(),
            "primaryCategory,secondaryCategory".to_string(),
        );

        let response = self
            .base
            .get(&format!("appInfos/{}", app_info_id), Some(params))
            .await?;

        let app_info = response.get("data").cloned().unwrap_or_default();
        let empty_vec = vec![];
        let included = response
            .get("included")
            .and_then(|i| i.as_array())
            .unwrap_or(&empty_vec);

        let mut category_lookup = HashMap::new();
        for item in included {
            if let (Some(item_type), Some(id), Some(attributes)) = (
                item.get("type").and_then(|t| t.as_str()),
                item.get("id").and_then(|i| i.as_str()),
                item.get("attributes"),
            ) {
                if item_type == "appCategories" {
                    category_lookup.insert(id.to_string(), attributes.clone());
                }
            }
        }

        let relationships = app_info.get("relationships").cloned().unwrap_or_default();
        let attributes = app_info.get("attributes").cloned().unwrap_or_default();

        let mut result = json!({
            "primaryCategory": null,
            "secondaryCategory": null,
            "primarySubcategoryOne": null,
            "primarySubcategoryTwo": null,
            "secondarySubcategoryOne": null,
            "secondarySubcategoryTwo": null
        });

        if let Some(primary_cat) = relationships
            .get("primaryCategory")
            .and_then(|pc| pc.get("data"))
        {
            if let Some(id) = primary_cat.get("id").and_then(|i| i.as_str()) {
                result["primaryCategory"] = json!({
                    "id": id,
                    "attributes": category_lookup.get(id).cloned().unwrap_or_default()
                });
            }
        }

        if let Some(secondary_cat) = relationships
            .get("secondaryCategory")
            .and_then(|sc| sc.get("data"))
        {
            if let Some(id) = secondary_cat.get("id").and_then(|i| i.as_str()) {
                result["secondaryCategory"] = json!({
                    "id": id,
                    "attributes": category_lookup.get(id).cloned().unwrap_or_default()
                });
            }
        }

        result["primarySubcategoryOne"] = attributes
            .get("primarySubcategoryOne")
            .cloned()
            .unwrap_or(Value::Null);
        result["primarySubcategoryTwo"] = attributes
            .get("primarySubcategoryTwo")
            .cloned()
            .unwrap_or(Value::Null);
        result["secondarySubcategoryOne"] = attributes
            .get("secondarySubcategoryOne")
            .cloned()
            .unwrap_or(Value::Null);
        result["secondarySubcategoryTwo"] = attributes
            .get("secondarySubcategoryTwo")
            .cloned()
            .unwrap_or(Value::Null);

        Ok(result)
    }

    pub async fn update_app_categories(
        &self,
        app_info_id: &str,
        primary_category_id: Option<&str>,
        secondary_category_id: Option<&str>,
        primary_subcategory_one: Option<&str>,
        primary_subcategory_two: Option<&str>,
        secondary_subcategory_one: Option<&str>,
        secondary_subcategory_two: Option<&str>,
    ) -> Result<Value, AppStoreConnectError> {
        let mut data = json!({
            "data": {
                "type": "appInfos",
                "id": app_info_id,
                "attributes": {},
                "relationships": {}
            }
        });

        let data_obj = data["data"].as_object_mut().unwrap();

        if let Some(subcategory) = primary_subcategory_one {
            data_obj["attributes"]["primarySubcategoryOne"] = json!(subcategory);
        }
        if let Some(subcategory) = primary_subcategory_two {
            data_obj["attributes"]["primarySubcategoryTwo"] = json!(subcategory);
        }
        if let Some(subcategory) = secondary_subcategory_one {
            data_obj["attributes"]["secondarySubcategoryOne"] = json!(subcategory);
        }
        if let Some(subcategory) = secondary_subcategory_two {
            data_obj["attributes"]["secondarySubcategoryTwo"] = json!(subcategory);
        }

        if let Some(category_id) = primary_category_id {
            data_obj["relationships"]["primaryCategory"] = json!({
                "data": {
                    "type": "appCategories",
                    "id": category_id
                }
            });
        }

        if let Some(category_id) = secondary_category_id {
            data_obj["relationships"]["secondaryCategory"] = json!({
                "data": {
                    "type": "appCategories",
                    "id": category_id
                }
            });
        }

        let response = self
            .base
            .patch(&format!("appInfos/{}", app_info_id), data)
            .await?;

        response
            .get("data")
            .cloned()
            .ok_or_else(|| AppStoreConnectError::Api {
                message: "Invalid response format".to_string(),
            })
    }

    pub async fn get_all_categories(
        &self,
        platform: Option<&str>,
    ) -> Result<Vec<Value>, AppStoreConnectError> {
        let mut params = HashMap::new();
        params.insert(
            "filter[platforms]".to_string(),
            platform.unwrap_or("IOS").to_string(),
        );
        params.insert("limit".to_string(), "200".to_string());

        let response = self.base.get("appCategories", Some(params)).await?;

        if let Some(data) = response.get("data").and_then(|d| d.as_array()) {
            Ok(data.clone())
        } else {
            Ok(Vec::new())
        }
    }

    pub async fn get_category_by_name(
        &self,
        category_name: &str,
        platform: Option<&str>,
    ) -> Result<Option<Value>, AppStoreConnectError> {
        let categories = self.get_all_categories(platform).await?;

        for category in categories {
            if let Some(display_name) = category
                .get("attributes")
                .and_then(|a| a.get("displayName"))
                .and_then(|n| n.as_str())
            {
                if display_name == category_name {
                    return Ok(Some(category));
                }
            }
        }

        Ok(None)
    }

    pub fn get_categories() -> HashMap<&'static str, &'static str> {
        let mut categories = HashMap::new();
        categories.insert("BOOKS", "Books");
        categories.insert("BUSINESS", "Business");
        categories.insert("DEVELOPER_TOOLS", "Developer Tools");
        categories.insert("EDUCATION", "Education");
        categories.insert("ENTERTAINMENT", "Entertainment");
        categories.insert("FINANCE", "Finance");
        categories.insert("FOOD_AND_DRINK", "Food & Drink");
        categories.insert("GAMES", "Games");
        categories.insert("GRAPHICS_AND_DESIGN", "Graphics & Design");
        categories.insert("HEALTH_AND_FITNESS", "Health & Fitness");
        categories.insert("LIFESTYLE", "Lifestyle");
        categories.insert("MAGAZINES_AND_NEWSPAPERS", "Magazines & Newspapers");
        categories.insert("MEDICAL", "Medical");
        categories.insert("MUSIC", "Music");
        categories.insert("NAVIGATION", "Navigation");
        categories.insert("NEWS", "News");
        categories.insert("PHOTO_AND_VIDEO", "Photo & Video");
        categories.insert("PRODUCTIVITY", "Productivity");
        categories.insert("REFERENCE", "Reference");
        categories.insert("SHOPPING", "Shopping");
        categories.insert("SOCIAL_NETWORKING", "Social Networking");
        categories.insert("SPORTS", "Sports");
        categories.insert("TRAVEL", "Travel");
        categories.insert("UTILITIES", "Utilities");
        categories.insert("WEATHER", "Weather");
        categories
    }

    pub fn get_game_subcategories() -> HashMap<&'static str, &'static str> {
        let mut subcategories = HashMap::new();
        subcategories.insert("ACTION", "Action");
        subcategories.insert("ADVENTURE", "Adventure");
        subcategories.insert("ARCADE", "Arcade");
        subcategories.insert("BOARD", "Board");
        subcategories.insert("CARD", "Card");
        subcategories.insert("CASINO", "Casino");
        subcategories.insert("CASUAL", "Casual");
        subcategories.insert("DICE", "Dice");
        subcategories.insert("EDUCATIONAL", "Educational");
        subcategories.insert("FAMILY", "Family");
        subcategories.insert("MUSIC", "Music");
        subcategories.insert("PUZZLE", "Puzzle");
        subcategories.insert("RACING", "Racing");
        subcategories.insert("ROLE_PLAYING", "Role Playing");
        subcategories.insert("SIMULATION", "Simulation");
        subcategories.insert("SPORTS", "Sports");
        subcategories.insert("STRATEGY", "Strategy");
        subcategories.insert("TRIVIA", "Trivia");
        subcategories.insert("WORD", "Word");
        subcategories
    }
}
